### DISCUSSION TITLE
Do I need to free the memory returned by `X509_get_ext_d2i()`?

### DISCUSSION THREAD

**Discussion by jasper-smit-servicenow** [openssl/openssl#24906] [Q&A] [answered]
Do I need to free the memory returned by `X509_get_ext_d2i()`? Examples seem to suggest that I need to free the `EXTENDED_KEY_USAGE *` with `EXTENDED_KEY_USAGE_free()`:  https://github.com/openssl/openssl/blob/d4700c0b237c05315e3bf14fc416abcbdfe51ff2/crypto/cmp/cmp_asn.c#L502

Also the documentation is not entirely clear to me. On https://www.openssl.org/docs/man1.1.1/man3/X509_get_ext_d2i.html states that:

> The function X509V3_add1_i2d() and its variants allocate X509_EXTENSION objects on STACK *x depending on flags. The X509_EXTENSION objects must be explicitly freed using X509_EXTENSION_free().

, but I am not sure if X509_get_ext_d2i() is considered a variant of X509V3_add1_i2d

What is the recommended approach here? 

Thanks,

**Comment by t8m** [comment, marked as answer]
Yes, you need to free the decoded extension object with an appropriate free function for the object decoded. The documentation could be improved. Would you like to submit a pull request with documentation improvement?
